py/replica_restart_policy_tests.py

```diff
@@ -0,0 +1,181 @@
+    logging.info("Current TFJob:\n %s", json.dumps(results, indent=2))
+
+    if replica_restart_policy == "Always":
+      tf_job_client.wait_for_replica_type_in_phases(
+        api_client, self.namespace, self.name, "PS", ["Running"])
+      first_start_time = tf_job_client.get_start_time_by_index(
+        api_client, self.namespace, self.name, "PS", 0)
+
+      tf_job_client.terminate_replicas(api_client, self.namespace, self.name,
+                                       "ps", 1, exit_code)
+
+      tf_job_client.wait_for_replica_type_in_phases(
+        api_client, self.namespace, self.name, "PS", ["Running"])
+      restart_time = tf_job_client.get_start_time_by_index(
+        api_client, self.namespace, self.name, "PS", 0)
+
+      logging.info("First start time: %s, restart time: %s",
+                   str(first_start_time), str(restart_time))
+
+      if restart_time <= first_start_time:
+        self.failure = "Job {0} in namespace {1} with restart policy Always failed test".format(
+          self.name, self.namespace)
+        logging.error(self.failure)
+        return
+
+    elif replica_restart_policy == "OnFailure":
+      tf_job_client.wait_for_replica_type_in_phases(
+        api_client, self.namespace, self.name, "PS", ["Running"])
+      first_start_time = tf_job_client.get_start_time_by_index(
+        api_client, self.namespace, self.name, "PS", 0)
+
+      tf_job_client.terminate_replicas(api_client, self.namespace, self.name,
+                                       "ps", 1, exit_code)
+
+      tf_job_client.wait_for_replica_type_in_phases(
+        api_client, self.namespace, self.name, "PS", ["Running"])
+      restart_time = tf_job_client.get_start_time_by_index(
+        api_client, self.namespace, self.name, "PS", 0)
+
+      logging.info("First start time: %s, restart time: %s",
+                   str(first_start_time), str(restart_time))
+
+      if restart_time <= first_start_time:
+        self.failure = "Job {0} in namespace {1} with restart policy OnFailure \
+          failed to restart the pod with exit_code 1".format(
+          self.name, self.namespace)
+        logging.error(self.failure)
+        return
+
+    elif replica_restart_policy == "Never":
+      tf_job_client.terminate_replicas(api_client, self.namespace, self.name,
+                                       "ps", 1, exit_code)
+      tf_job_client.wait_for_replica_type_in_phases(api_client, self.namespace,
+                                                    self.name, "PS", ["Failed"])
+
+    elif replica_restart_policy == "ExitCode" and exit_code == 1:
+      tf_job_client.terminate_replicas(api_client, self.namespace, self.name,
+                                       "ps", 1, exit_code)
+      tf_job_client.wait_for_replica_type_in_phases(api_client, self.namespace,
+                                                    self.name, "PS", ["Failed"])
+
+    else:
+      tf_job_client.wait_for_replica_type_in_phases(
```

Comment:
Consider refactoring the code to verify start time to another helper method. For example:

def terminate_and_verify_start_time(replica_type, replica_index, exit_code, expect_restart)
# if expect_restart is true, check that the second restart time is after the first.
# if expect_restart is false, check that the restart time has not changed.